src/flint/test/test_all.py

```diff
@@ -3229,6 +3229,38 @@ def quick_poly():
             assert raises(lambda: quick_poly().gcd(None), TypeError)
             assert raises(lambda: quick_poly().gcd(P(ctx=ctx1)), IncompatibleContextError)
 
+        x0, x1 = ctx.gens()
+        assert (2*x0**3 + 4*x0**2 + 6*x0).term_content() == x0 if is_field else 2*x0
+        assert (3*x0**2*x1 + 6*x0*x1**2 + 9*x1**3).term_content() == x1 if is_field else 3*x1
+
+        assert (x0**2 - 1).resultant(x0**2 - 2*x0 + 1, 'x0') == 0
+        assert (x0**2 + x1**2 - 1).resultant(x0 - x1, 'x0') == 2*x1**2 - 1
+
+        assert (x0**3 - 6*x0**2 + 11*x0 - 6).discriminant('x0') == 4
+        assert (x0**2 + 4*x0*x1 + 4*x1**2 - 1).discriminant('x0') == 4
+
+        f1 = 3*x0**2*x1**2 + 6*x0*x1**2 + 9*x1**2
+        res, stride = f1.deflation()
+        assert res == 3*x0**2 + 6*x0 + 9
+        assert tuple(stride) == (1, 0)
+
+        g1 = ((x0**2 + x1**2)**3 + (x0**2 + x1**2)**2 + 1)
+        res, stride = g1.deflation()
+        assert res == x0**3 + 3*x0**2*x1 + x0**2 + 3*x0*x1**2 + 2*x0*x1 + x1**3 + x1**2 + 1
+        assert tuple(stride) == (2, 2)
+
+        for p in [f1, g1]:
+            n, m = p.deflation_monom()
+            assert m * p.deflate(n).inflate(n) == p
+
+            n, i = p.deflation_index()
+            m = ctx.term(exp_vec=i)
+            assert p.deflate(n).inflate(n) * m == p
```

Comment:
These don't quite seem right. It should be:
```python
assert (p / m).deflate(n).inflate(n) * m = p
```
In other words `.deflate(n)` and `.inflate(n)` should be direct inverses.

In the first case I was expecting that `deflation_monom()` returns a deflated polynomial i.e.:
```python
pd, n, m = p.deflation_monom()
assert m * pd.inflate(n)
```
There is less value in having two functions for this if the only difference is whether `m` is a list of integers or a monomial since, as demonstrated here, it is not too difficult to convert between those forms.

Reply:
> In other words .deflate(n) and .inflate(n) should be direct inverses.

There is something more wrong with the way I've implemented these functions then, but this doesn't seem possible by using a zero-shift vector. 

```python
In [47]: x.deflation()
Out[47]: (1, [0, 0])

In [48]: x.deflate([0, 0]).inflate([0, 0])
Out[48]: 1
```

When the exponent of a variable matches in all instances the shift vector then contains all the information, it's then discarded.

Reply:
`deflation_monom` turns the shift vector into the monomial so I don't think it's an issue there. Let me re-review your pervious comments. I think I've misunderstood something

Reply:
I think both the shift and stride vectors are strictly required. The multivariate deflation does more than it's univariate counter part because the univariate form does not allow mapping x^n -> x^0 while the multivariate form does.

Reply:
Can you not set the shifts to zero or is that the bug that you found?

I was imagining something like this:
```python
shifts, strides = foo_mpoly_deflation(...)

for i in range(n):
    g = gcd(shifts[i], strides[i])
    if g != 1: # maybe this check not needed?
        strides[i] = g
    shifts[i] = 0

p = foo_mpoly_deflate(shifts, strides, ...)
return p, strides
```

Reply:
We should also check what the `gr_poly` and `gr_mpoly` functions do for things like this but it seems that they don't define this yet...

Reply:
> Can you not set the shifts to zero or is that the bug that you found?

I believe that was unrelated but might be the reason why that went undiscovered for 6 years. The current implementation sets the shift vector to zeros. But the returned stride vector is only correct when used with the shift vector.

> I was imagining something like this:
> 
> ```python
> shifts, strides = foo_mpoly_deflation(...)
> 
> for i in range(n):
>     g = gcd(shifts[i], strides[i])
>     if g != 1: # maybe this check not needed?
>         strides[i] = g
>     shifts[i] = 0
> 
> p = foo_mpoly_deflate(shifts, strides, ...)
> return p, strides
> ```

I've give that a shot

Reply:
This should be resolved now